src/backend/distributed/commands/dependencies.c

```diff
@@ -0,0 +1,164 @@
+
+
+/*
+ * EnsureDependenciesExists finds all the dependencies that we can distribute and makes
+ * sure these are available on all workers. If not available they will be created on the
+ * workers via a separate session that will be committed directly so that the objects are
+ * visible to potentially multiple sessions creating the shards.
+ */
+void
+EnsureDependenciesExistsOnAllNodes(const ObjectAddress *target)
+{
+	const uint64 connectionFlag = FORCE_NEW_CONNECTION;
+	ListCell *dependencyCell = NULL;
+
+	List *dependencies = NIL;
+	List *connections = NULL;
+	ListCell *connectionCell = NULL;
+
+	/* collect all dependencies in creation order and get their ddl commands */
+	GetDependenciesForObject(target, &dependencies);
+
+	/* create all dependencies on all nodes and mark them as distributed */
+	foreach(dependencyCell, dependencies)
+	{
+		ObjectAddress *dependency = (ObjectAddress *) lfirst(dependencyCell);
+		List *ddlCommands = GetDependencyCreateDDLCommands(dependency);
+
+		if (list_length(ddlCommands) <= 0)
+		{
+			continue;
+		}
+
+		/* initialize connections on first commands to execute */
+		if (connections == NULL)
+		{
+			/* first command to be executed connect to nodes */
+			List *workerNodeList = ActivePrimaryNodeList();
+			ListCell *workerNodeCell = NULL;
+
+			if (list_length(workerNodeList) <= 0)
+			{
+				/* no nodes to execute on, we can break out */
+				break;
+			}
+
+			foreach(workerNodeCell, workerNodeList)
+			{
+				WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+				MultiConnection *connection = NULL;
+
+				char *nodeName = workerNode->workerName;
+				uint32 nodePort = workerNode->workerPort;
+
+				connection = GetNodeUserDatabaseConnection(connectionFlag, nodeName,
+														   nodePort,
+														   CitusExtensionOwnerName(),
+														   NULL);
+
+				connections = lappend(connections, connection);
+			}
+		}
+
+		/* create dependency on all worker nodes*/
+		foreach(connectionCell, connections)
+		{
+			MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+			ExecuteCriticalRemoteCommandList(connection, ddlCommands);
+		}
+
+		/* mark the object as distributed in this transaction */
+		recordObjectDistributed(dependency);
+	}
+
+	foreach(connectionCell, connections)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+		CloseConnection(connection);
+	}
+}
+
+
+static List *
+GetDependencyCreateDDLCommands(const ObjectAddress *dependency)
+{
+	switch (getObjectClass(dependency))
+	{
+		case OCLASS_SCHEMA:
+		{
+			const char *schemaDDLCommand = CreateSchemaDDLCommand(dependency->objectId);
+
+			if (schemaDDLCommand == NULL)
+			{
+				/* no schema to create */
+				return NIL;
+			}
+
+			return list_make1((void *) schemaDDLCommand);
+		}
+
+		default:
+		{
+			return NIL;
+		}
+	}
+}
+
+
+void
+ReplicateAllDependenciesToNode(const char *nodeName, int nodePort)
```

Comment:
Document function
Potential to DRY with above function
Requires markObjectDistributed to be either idempotent or after invocation to create on all nodes, or both